File: dsa-old/data_structures/Queue/queue.cpp
#include <iostream>

#include <climits>

using namespace std;

class Queue {
    private: int front;
    int rear;
    int * arr;
    int size;

    public: Queue(int size) {
            cout << "Constructor called!" << endl;
            front = -1;
            rear = -1;
            this -> size = size;
            arr = new int[size]; // array size determined at run-time
        }

        ~Queue() {
            cout << "Destructor called!" << endl;
            delete[] arr;
        }

    bool isFull();
    bool isEmpty();
    void enqueue(int);
    int dequeue();
    void display();
};

bool Queue::isFull() {
    if (rear == size - 1) {
        return 1;
    }
    return 0;
}

bool Queue::isEmpty() {
    if (rear == -1 || front == rear) {
        return 1;
    }
    return 0;
}

void Queue::enqueue(int data) {
    if (!isFull()) {
        rear++;
        arr[rear] = data;
        return;
    }
    cout << "Queue is full!" << endl;
}

int Queue::dequeue() {
    if (!isEmpty()) {
        front++;
        return arr[front];
    }
    cout << "Queue is empty!" << endl;
    return INT_MIN;
}

void Queue::display() {
    if (!isEmpty()) {
        cout << "The data are: ";
        for (int i = front + 1; i <= rear; i++) {
            cout << arr[i] << " ";
        }
        cout << endl;
        return;
    }
    cout << "Queue is empty!" << endl;
}

void queueOperations(Queue & );

int main() {
    Queue q1(5);
    queueOperations(q1);
}

void queueOperations(Queue & myObj) {
    int choice;
    int popped;

    cout << "1- enqueue, 2-dequeue, 3- Display, 4- EXIT" << endl;
    do {
        cout << "Enter choice: ";
        cin >> choice;

        switch (choice) {
        case 1:
            int ele;
            cout << "Enter data to be inserted: ";
            cin >> ele;
            myObj.enqueue(ele);
            break;
        case 2:
            popped = myObj.dequeue();
            if (popped != INT_MIN) {
                cout << "Popped: " << popped << endl;
            } else {
                cout << "Queue is empty!" << endl;
            }
            break;
        case 3:
            myObj.display();
            break;
        case 4:
            // when we used exit(0), destructor was not being called because exit(0) directly exits the program. Using 'return' will return to the main function and thus, destructor will be called
            return;
        default:
            cout << "Wrong Choice!" << endl;
            break;
        }
    } while (1);
}

File: dsa/data-structures/Queue.java
import java.util.Scanner;

class QueueImplementation {
  // front->[1][2][3][4][5][6]<-rear
  private int front; // at the start, used for dequeue
  private int rear; // at the end, used for enqueue
  private int size;
  private int[] queue;

  QueueImplementation(int size) {
    this.size = size;
    queue = new int[size];
    front = -1;
    rear = -1;
  }

  void display() {
    if (isEmpty()) {
      System.out.println("Queue is empty");
      return;
    }
    
    for (int i = front; i <= rear; i++) {
      System.out.print(queue[i] + " ");
    }
    System.out.println();
  }

  boolean isEmpty() {
    return front == -1;
  }

  boolean isFull() {
    return rear == size - 1;
  }

  void enqueue(int val) {
    if (isFull()) {
      System.out.println("Queue is already full");
      return;  
    }

    if (front == -1) front = 0;
    queue[++rear] = val;
  }

  int dequeue() {
    if (!isEmpty()) return queue[front++];
    System.out.println("Queue is already empty");
    return -1;
  }
}

class Queue {
  public static void main(String[] args) {
    QueueImplementation queue = new QueueImplementation(5);

    System.out.println("1- Enqueue, 2- Dequeue, 3- Display, 4- Exit");
    while(true) {
      System.out.println();
      Scanner sc = new Scanner(System.in);
      System.out.print("Enter choice: ");
      int choice = sc.nextInt();
      
      switch(choice) {
        case 1:
          System.out.print("Enter val: ");
          int val = sc.nextInt();
          queue.enqueue(val);
          break;

        case 2:
          int poppedVal = queue.dequeue();
          if (poppedVal != -1) System.out.println("Popped val is: " + poppedVal);
          break;

        case 3:
          queue.display();
          break;

        case 4:
          System.exit(0);
          break;

        default:
          System.out.println("Wrong choice!");
          break;
      }
    }
  }
}
